vendor/github.com/thanos-io/objstore/providers/s3/s3.go

```diff
@@ -510,6 +514,7 @@ func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
 			ServerSideEncryption: sse,
 			UserMetadata:         userMetadata,
 			StorageClass:         b.storageClass,
+			SendContentMd5:       b.sendContentMd5,
```

Comment:
We should double-check what's the implication of setting `SendContentMd5` to `true` by default. Does it mean we will use CPU to compute MD5 of uploaded objects?

Reply:
> Does it mean we will use CPU to compute MD5 of uploaded objects?

I believe so (I have not stepped through or ran tests to confirm). In minio's code there are comments like [this](https://github.com/minio/minio-go/blob/e0b68385002c0f63d5ea17f94eacc9043acdef24/api-put-object-streaming.go#L337): `// CRC32C is ~50% faster on AMD64 @ 30GB/s`.

The motivation for the [objstore PR](https://github.com/thanos-io/objstore/pull/85) seems to have been compatibility by default with Backblaze B2. On our side the currently open https://github.com/grafana/mimir/pull/6592 seems to be related to the same issue (but with a more indirect solution).

I think exposing the option makes sense for us now, but choosing the default in this situation is tricky since both options have downsides. `true` adds a hidden cost to everyone (it's not [required](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html?icmpid=docs_amazons3_console#checking-object-integrity-md5)), `false` adds a visible and correctable cost to some. Looking even further ahead, there's trailing headers (https://github.com/thanos-io/objstore/issues/57) to keep in mind.

Seems like a tough call. I'm slightly more in favor of `false`.

Reply:
My understanding is that it was `false` until now, so I'd just keep current behaviour, and perhaps expose an option to change it by people who need it. WDYT?

Reply:
[Minio](https://min.io/docs/minio/linux/developers/go/API.html) warns that `SendContentMd5` will cause higher memory usage "because of in-memory md5sum calculation". Also the code comment (linked to by Andy) states 50% worse perf with MD5 over CRC32. I would default to `false`, as that should be the behaviour before this PR, but we could expose the option.

Reply:
> My understanding is that it was false until now, so I'd just keep current behaviour

>  I would default to false, as that should be the behaviour before this PR

Seems like there's agreement on `false` then :+1:

As an aside my understanding was that the past behavior had _already_ changed in some way like:
1. Minio not using MD5 or checksum when we didn't set `true` to send an MD5
2. Minio starts using checksums when not sending an MD5 <-- breakage for some users happens here
3. Now

Oddly the [minio issue](https://www.github.com/minio/minio-go/issues/1791) related to this was from quite a while ago. The Thanos bump pointed to in the objstore PR doesn't include a minio upgrade at all according to the `go.sum`. I feel like there's some confusion somewhere regarding what change actually caused this to surface more recently, but even if that archaeology was performed I don't think it changes anything for us.

Reply:
> Minio starts using checksums when not sending an MD5 <-- breakage for some users happens here

Yeah that was also my understanding from digging into the backstory (CRC checksum + an Amazon checksum header). It's confusing, and not well communicated :disappointed:

Reply:
Added `-<prefix>.s3.send-content-md5` defaulting to false.

Reply:
Do we need to support this MD5 thing at all in Mimir?